iceoryx_posh/source/popo/wait_set.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/condition_variable_liveliness.hpp"
+
+namespace iox
+{
+namespace popo
+{
+WaitSet::WaitSet(cxx::not_null<ConditionVariableData* const> condVarDataPtr) noexcept
+    : m_conditionVariableDataPtr(condVarDataPtr)
+    , m_conditionVariableWaiter(m_conditionVariableDataPtr)
+{
+}
+
+WaitSet::~WaitSet() noexcept
+{
+    // Notify all conditions in the vector that the condition variable data will be destroyed
+    detachAllConditions();
+    /// @todo Notify RouDi that the condition variable data shall be destroyed
+}
+
+cxx::expected<WaitSetError> WaitSet::attachCondition(Condition& condition) noexcept
+{
+    if (!condition.isConditionVariableAttached())
+    {
+        if (condition.attachConditionVariableIntern(m_conditionVariableDataPtr))
+        {
+            if (!m_conditionVector.push_back(&condition))
+            {
+                return cxx::error<WaitSetError>(WaitSetError::CONDITION_VECTOR_OVERFLOW);
+            }
+
+            ConditionVariableLiveliness condVarLiveliness{m_conditionVariableDataPtr};
+            condVarLiveliness.announce();
+            return iox::cxx::success<>();
+        }
+    }
+    return cxx::error<WaitSetError>(WaitSetError::CONDITION_VARIABLE_ALREADY_SET);
+}
+
+bool WaitSet::detachCondition(Condition& condition) noexcept
+{
+    if (condition.isConditionVariableAttached())
+    {
+        if (!condition.detachConditionVariable())
+        {
+            errorHandler(Error::kPOPO__WAITSET_COULD_NOT_DETACH_CONDITION, nullptr, ErrorLevel::FATAL);
+        }
+
+        for (auto& currentCondition : m_conditionVector)
+        {
+            if (currentCondition == &condition)
+            {
+                ConditionVariableLiveliness condVarLiveliness{m_conditionVariableDataPtr};
+                condVarLiveliness.recall();
+                m_conditionVector.erase(&currentCondition);
+                return true;
+            }
+        }
+    }
```

Comment:
Similar situation here. You can attach and detach a condition variable without the need to have space in your vector or the need for the condition to be known by your waitset. Is this intended behavior?
Do you have tests like detach a condition variable that was never attached?

Reply:
Yeah, refactored it. Push is now done before attach. `DetachConditionNotInListResultsInFailure` is there.